hopsworks-api/src/main/java/io/hops/hopsworks/api/jwt/JWTResource.java

```diff
@@ -64,10 +77,55 @@ public Response createToken(JWTRequestDTO jWTRequestDTO) throws NoSuchAlgorithmE
   @Path("/renew")
   public Response renewToken(JsonWebTokenDTO jsonWebTokenDTO) throws SigningKeyNotFoundException, NotRenewableException,
       InvalidationException {
-    JWTResponseDTO jWTResponseDTO = jWTHelper.renewToken(jsonWebTokenDTO);
+    JWTResponseDTO jWTResponseDTO = jWTHelper.renewToken(jsonWebTokenDTO, true, new HashMap<>(3));
     return Response.ok().entity(jWTResponseDTO).build();
   }
 
+  @PUT
+  @Path("/renew/service")
+  public Response renewServiceToken(JsonWebTokenDTO jwt, @Context HttpServletRequest request)
+      throws HopsSecurityException {
+    // This token should be the one-time renewal token
+    String token = jWTHelper.getAuthToken(request);
+    Users user = jWTHelper.getUserPrincipal(request);
+    if (user == null) {
+      DecodedJWT decodedJWT = JWT.decode(token);
+      throw new HopsSecurityException(RESTCodes.SecurityErrorCode.NOT_RENEWABLE_TOKEN, Level.FINE,
+          "User not found associated with that JWT", "Could not find user in the database associated with JWT "
+          + decodedJWT.getId());
+    }
+    try {
+      ServiceJWTDTO renewedTokens = jWTHelper.renewServiceToken(jwt, token, user, request.getRemoteHost());
+      return Response.ok().entity(renewedTokens).build();
+    } catch (JWTException | NoSuchAlgorithmException ex) {
+      throw new HopsSecurityException(RESTCodes.SecurityErrorCode.NOT_RENEWABLE_TOKEN, Level.WARNING,
+          "Could not renew service JWT", "Could not renew service JWT for " + request.getRemoteHost());
+    }
+  }
+  
+  @DELETE
+  @Path("/invalidate/service/{token}")
+  public Response invalidateServiceToken(@PathParam("token") String token, @Context HttpServletRequest request)
+    throws HopsSecurityException {
+    DecodedJWT jwt2invalidate = JWT.decode(token);
+    
+    Users user = jWTHelper.getUserPrincipal(request);
+    if (user == null) {
+      throw new HopsSecurityException(RESTCodes.SecurityErrorCode.INVALIDATION_ERROR, Level.FINE,
+          "Could not find registered user", "Could not find registered user associated with JWT "
+            + jwt2invalidate.getId());
+    }
+    if (!user.getUsername().equals(jwt2invalidate.getSubject())) {
+      throw new HopsSecurityException(RESTCodes.SecurityErrorCode.INVALIDATION_ERROR, Level.FINE,
+          "Tried to invalidate token with different subject",
+          "User " + user.getUsername() + " tried to invalidate token with Subject " + jwt2invalidate.getSubject());
+    }
+    
+    jWTHelper.invalidateServiceToken(token);
+    
+    return Response.ok().build();
+  }
+  
   @POST
   @Path("/invalidate")
```

Comment:
Same here. I know you didn't touch this code, but as you are changing both this class and Hops it would be nice if we can have a coherent, REST api.
What about just `DELETE` on the `/` or a POST/PUT with `?action=invalidate`?
@tkakantousis any suggestion?

Reply:
Yes both make sense, the first approach is preferred. 
So `DELETE /` is invalidating the token? Then in a previous comment I see `DELETE -> /service/{token} -> Invalidate a token` which is supposed to do the same thing?

Reply:
@tkakantousis There are two kinds of tokens. JWTs issued for applications and JWTs issued for services which require different handling. Currently `/invalidate` is used for invalidating applications' tokens and `/invalidate/service/{token}` for services' tokens.

Reply:
then it should be `DELETE /application` and `DELETE /service` or `DELETE ?type=application|service` or something like that. The first approach would be preferable